pkg/jobWorker/job.go

```diff
@@ -0,0 +1,307 @@
+		return fmt.Errorf("could not add resources into controller:%s, %v", ns.MEMORY_HIGH_File, err)
+	}
+	//if err = ns.AddResourceControl(cgroupName, ns.IO_WEIGHT_File, strconv.FormatInt(job.config.IOBytesPerSecond, 10)); err != nil {
+	//	return fmt.Errorf("could not add resources into controller:%s, %v", ns.IO_WEIGHT_File, err)
+	//}
+
+	//provide the file descriptor to cmd.Run so that it can add the new PID to the control group
+	if err = ns.AddProcess(cgroupName, cmd); err != nil {
+		log.Printf("Error AddProcess /proc - %s\n", err)
+		os.Exit(1)
+	}
+
+	if err = ns.MountProc(); err != nil {
+		log.Printf("Error mounting /proc - %s\n", err)
+		os.Exit(1)
+	}
+
+	log.Printf("starting job:%s, cmd:%s", job, cmd.String())
+	if err = cmd.Start(); err != nil {
+		return fmt.Errorf("error starting command: %w", err)
+	}
+	job.cmd = cmd
+	job.status.State = JOB_STATUS_RUNNING
+
+	// run the command in a Goroutine so that Start can return immediately
+	go func() {
+		// Use cmd.Process.Wait() instead of cmd.Wait() since cmd.Wait() is not thread safe
+		// and we do not want to hold the mutex while waiting for the process to exit.
+		// So instead we use cmd.Process.Wait() and store the result in j.processState to mimic what
+		// cmd.Wait would do.
+		// This prevents concurrency issues when a user calls Start(), the command quickly exits (updating the
+		// process state), and the user invokes Status().
+		processState, err := job.cmd.Process.Wait()
+
+		job.mutex.Lock()
+		defer job.mutex.Unlock()
+
+		job.processState = processState
+		job.status.ExitCode = job.processState.ExitCode()
+		if job.status.State == JOB_STATUS_RUNNING {
+			job.status.State = JOB_STATUS_COMPLETED
+		}
+
+		if err != nil {
+			job.status.ExitReason = job.status.ExitReason + fmt.Sprintf("error running command: %s\n", err)
+		}
+
+		if err == nil && !job.processState.Success() {
+			job.status.ExitReason = job.status.ExitReason + fmt.Sprintf("error running command: %s\n", &exec.ExitError{ProcessState: job.processState})
+		}
+
+		// close the output, so that any readers of the output know the process has exited and will no longer
+		// block waiting for new output
+		if err = job.output.Close(); err != nil {
+			job.status.ExitReason = job.status.ExitReason + fmt.Sprintf("error closing output: %s\n", err)
+		}
+
+		// do not close the cgroup.procs file until after the process has exited
+		if err = ns.DeleteCGroup(cgroupName); err != nil {
+			job.status.ExitReason = job.status.ExitReason + fmt.Sprintf("error closing cgroup: %s\n", err)
+		}
+
+		if err = ns.UnmountProc(); err != nil {
+			log.Printf("error unmounting /proc - %s\n", err)
+		}
+	}()
+
+	return nil
+}
+
+// Status returns the current Status of the Job.
+func (job *Job) Status() *JobStatus {
+	job.mutex.Lock()
+	defer job.mutex.Unlock()
+	log.Printf("get job status:%s", job)
+	return job.status
```

Comment:
This could lead to a potential race condition when using the library:

```
status := job.Status()
if status.State == JOB_STATUS_RUNNING { 
}
```

Since status.State refers to the same memory address, it could be modified by other functions (e.g., job.status.State = JOB_STATUS_TERMINATED) without proper locking, leading to unpredictable behavior during job handling.

Reply:
@smallinsky good point. The best option in this case create set of internal flags and based on them return appropriate status.

<img width="668" alt="Screenshot 2024-09-04 at 1 49 59 PM" src="https://github.com/user-attachments/assets/1d5c66c9-d68b-46f4-98cf-31a19cb5fd79">